File: 6_kyu/Take_a_Ten_Minutes_Walk.lua
return {
  -- Check the walk plan.
  -- The walk plan is an array of steps in the direction n,s,w or e.
  -- The walk must contain exact 10 steps.
  -- The walk must return to the start point.
  is_valid_walk = function(walk)
    if walk == nil or #walk ~= 10 then
      -- Walk has not correct length of 10 steps.
      return false
    end
    local counter = {n=0, s=0,e=0,w=0}
    for _,direction in pairs(walk) do
      if counter[direction] == nul then
        -- Invalid direction code in the walk plan
        return false
      end
      counter[direction] = counter[direction] + 1
    end
    -- the walk must return to the start point.
    return counter.s == counter.n and counter.e == counter.w
  end
}

File: 6_kyu/Take_a_Ten_Minutes_Walk.go
package kata

func IsValidWalk(walk []rune) bool {
	// Quick check: 10 steps
	if len(walk) != 10 {
		return false
	}
	// Check: the work goes back to the start position
	positionNS := 0
	positionEW := 0
	for _, r := range walk {
		switch r {
		case 'n':
			positionNS += 1
		case 's':
			positionNS -= 1
		case 'e':
			positionEW += 1
		case 'w':
			positionEW -= 1
		default:
			return false // Invalid direction
		}
	}
	return positionNS == 0 && positionEW == 0
}
